pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,233 @@
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	list, _ := c.workStatusLister.ByNamespace("").List(labels.Everything())
+	for _, obj := range list {
+		wsRef, err := runtimeObjectToWorkStatusRef(obj)
+		if err != nil {
+			return err
+		}
+		sourceObjID := wsRef.SourceObjectIdentifier
+		if sourceObjID != wObjID {
+			continue
+		}
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		logger.V(4).Info("Updating singleton status", "workstatus", wsRef.Name, "wecName", wsRef.WECName)
+		return updateObjectStatus(ctx, &wsRef.SourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+	}
+	return nil
+}
+
+// reconcileSingletonByWs decides sync or not for the corresponding workload object,
+// then maintains its status based on the decision.
+func (c *Controller) reconcileSingletonByWs(ctx context.Context, ref singletonWorkStatusRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton status due to workstatus changes", "name", string(ref.Name))
+
+	wObjID := ref.SourceObjectIdentifier
+	wObjGVR := wObjID.GVR()
+	lister, found := c.listers.Get(wObjGVR)
+	if !found {
+		return fmt.Errorf("could not get lister for gvr: %s", wObjGVR)
+	}
+
+	var wObj runtime.Object
+	var err error
+	if wObjID.ObjectName.Namespace != "" {
+		wObj, err = lister.ByNamespace(wObjID.ObjectName.Namespace).Get(wObjID.ObjectName.Name)
+	} else {
+		wObj, err = lister.Get(wObjID.ObjectName.Name)
+	}
+	if err != nil {
+		return err
+	}
+
+	labels := wObj.(metav1.Object).GetLabels()
+	if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; !ok {
+		return nil
+	} else if v == util.BindingPolicyLabelSingletonStatusValueUnset {
+		return c.reconcileSingletonWObj(ctx, wObjID, false)
+	}
+
+	sync := map[util.ObjectIdentifier]bool{wObjID: false}
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, bdg := range allBdgs.Items {
+		if err = c.checkSingletonWorkloadInBinding(ctx, bdg, sync, false); err != nil {
+			return err
+		}
+	}
+
+	if !sync[wObjID] {
+		logger.V(4).Info("Singleton workload object should not have status synced, cleaning",
+			"resource", ref.SourceObjectIdentifier.Resource, "objectName", ref.SourceObjectIdentifier.ObjectName)
+		return c.reconcileSingletonWObj(ctx, wObjID, false)
+	} else {
+		logger.V(4).Info("Singleton workload object should have status synced, updating",
+			"resource", ref.SourceObjectIdentifier.Resource, "objectName", ref.SourceObjectIdentifier.ObjectName)
+		wsObj, err := c.workStatusLister.ByNamespace(ref.WECName).Get(ref.Name)
+		if err != nil {
+			return err
+		}
+		status, err := util.GetWorkStatusStatus(wsObj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		return updateObjectStatus(ctx, &wObjID, status, c.listers, c.wdsDynClient)
```

Comment:
There are two copies of this code. The other copy is in reconcileSingletonWObj. The other copy includes debug logging of this action, which is a good thing.

Reply:
Addressed in #2409.